### PR TITLE
HAWQ-316. Rework SSE42 implementation and runtime logic to be more similar to PostgreSQL9.5

### PR DESCRIPTION
**Pull request by liming01**
The problem already fixed on GPDB, 
https://github.com/greenplum-db/gpdb/pull/31

Here just merge the fix into hawq.

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -3832,7 +3833,8 @@ RecordIsValid(XLogRecord *record, XLogRecPtr recptr, int emode)
 	 */
 
 	/* First the rmgr data */
-	crc = crc32c(crc32cInit(), XLogRecGetData(record), len);
+    INIT_CRC32C(crc);
```
Comment (wangzw) [inline src/backend/access/transam/xlog.c]: please fix indent issue.

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -4914,10 +4916,11 @@ WriteControlFile(void)
 	StrNCpy(ControlFile->lc_ctype, localeptr, LOCALE_NAME_BUFLEN);
 
 	/* Contents are protected with a CRC */
-	ControlFile->crc = crc32c(crc32cInit(),
+    INIT_CRC32C(ControlFile->crc);
```
Comment (wangzw) [inline src/backend/access/transam/xlog.c]: please fix indent issue.

File: src/bin/gpupgrade/gpmodcatversion.c [hunk 3]
```diff
@@ -174,19 +174,20 @@ main(int argc, char *argv[])
 	}
 
 	/* Check the CRC. */
-	crc = crc32c(crc32cInit(), &ControlFile, offsetof(ControlFileData, crc));
-	crc32cFinish(crc);
+    INIT_CRC32C(crc);
```
Comment (wangzw) [inline src/bin/gpupgrade/gpmodcatversion.c]: please fix indent issue.

File: src/bin/gpupgrade/gpmodcatversion.c [hunk 4]
```diff
@@ -224,14 +225,16 @@ main(int argc, char *argv[])
 		ControlFile.catalog_version_no = tover;
 
 		/* recalcualte the CRC. */
-		crc = crc32c(crc32cInit(), &ControlFile, offsetof(ControlFileData, crc));
-		crc32cFinish(crc);
+        INIT_CRC32C(crc);
+    	COMP_CRC32C(crc, &ToControlFile, offsetof(ControlFileData, crc));
+    	FIN_CRC32C(crc);
+
```
Comment (wangzw) [inline src/bin/gpupgrade/gpmodcatversion.c]: please fix indent issue.

File: src/bin/pg_controldata/pg_controldata.c [hunk 5]
```diff
@@ -129,23 +129,24 @@ main(int argc, char *argv[])
 	close(fd);
 
 	/* Check the CRC. */
-	crc = crc32c(crc32cInit(), &ControlFile, offsetof(ControlFileData, crc));
-	crc32cFinish(crc);
+    INIT_CRC32C(crc);
```
Comment (wangzw) [inline src/bin/pg_controldata/pg_controldata.c]: please fix indent issue.

**Comment by wangzw** [comment]
The patch is generally OK but I have some concerns.

1) Do we have test case for CRC calculation?
2) By default the checksum verify for interconnect module is disable, but we do not have test case for CRC enable interconnect. It make code change unsafe.

**Comment by cwelton** [comment]
What will it take to get this moving again?  Can we merge it or close it out?

**Comment by ivannovick** [comment]
@liming01 how is this being tested?  Can you summarize the test cases for this change?

**Comment by liming01** [comment]
1) Not specific test case for this changes.
2) By default, although HAWQ disable the checksum verification for interconnect module, the catalog stored in local disk will use it.

**Comment by cwelton** [comment]
This PR has now been open for over 30 days.  

I see that HAWQ-316 has been marked resolved, did you simply forget to close the PR?

**Comment by liming01** [comment]
Not yet, I haven't gotten approval to merge it.
I will merge it ASAP when getting approval. Thanks.

**Comment by changleicn** [comment]
looks good!

**Comment by linwen** [comment]
+1
